### ISSUE TITLE
thanos-compact should exit on critical error

### ISSUE THREAD

**Issue by belm0** [thanos-io/thanos#2887]
**What happened**:

thanos-compact has a disk full or encounters overlapping blocks, and enters "halt" state but keeps running

**What you expected to happen**:

thanos-compact should exit.  A flapping job has much higher chance of being noticed in typical deployments.

we should all have wonderful monitoring of server logs and metrics, but usually don't...

real world example:  for several weeks we didn't notice that thanos-compact was halted due to overlapping blocks uploaded by a rouge prometheus+thanos cluster on the same replica name and store.  Hundreds of overlapping blocks were added during this time.

**Comment by belm0** [comment]
> we should all have wonderful monitoring of server logs and metrics, but usually don't...

E.g. even the thanos-compact Grafana dashboard does not surface `thanos_compactor_halted`, which is probably the most important signal to show users given the current behavior to continue running.

**Comment by GiedriusS** [comment]
> > we should all have wonderful monitoring of server logs and metrics, but usually don't...
> 
> E.g. even the thanos-compact Grafana dashboard does not surface `thanos_compactor_halted`, which is probably the most important signal to show users given the current behavior to continue running.

Then maybe we should fix that instead? :P I think the behaviour of halting on such an error is _very_ important because it let's us inspect the state of what is on disk before retrying everything again.

**Comment by belm0** [comment]
> I think the behaviour of halting on such an error is very important because it let's us inspect the state of what is on disk before retrying everything again.

@GiedriusS How many users will actually do that?  It sounds like a rare thing that a thanos developer might do.

Please at least consider a flag to control whether the server exits or not.

**Comment by midnightconman** [comment]
I agree with @belm0, thanos compact should fatal on an unrecoverable error (like a "no space left on disk" error). If you don't fatal, there is no possibility of automated recovery.

On the flip side, if we were to fatal on out of space issues, we could resize the PVC in kubernetes we have for the compactor and on its next loop it would recover. Or if we didn't have a PVC and the workload shifted to another host, that could also result in automatic recovery.

**Comment by bwplotka** [comment]
So if not halting is what interests you there is hidden `--debug.halt-on-error=false` you can set (:

**Comment by belm0** [comment]
> there is hidden --debug.halt-on-error=false you can set 

thank you, that sounds like it will work

But I have to say again that the default behavior seems backwards.    When there is a critical error a server should abort.  That is what's expected and useful for users.

I understand the use case where you want to halt the process and do a post-mortem, but that is particular to developers of the server itself, and not the common case.

https://github.com/thanos-io/thanos/blob/2ea2c2b7c4e6ac157925a3a9590dd5ac47bb1ca9/cmd/thanos/compact.go#L367-L377

**Comment by bwplotka** [comment]
We can definitely discuss this - in fast let's reopen this to do so (: 

So far the major reason behind this is that:

* Compactor does not have to be 100% available. It's like periodic batch job rather.
* For most of the errors restart does not help in any form. e.g corrupted block, overlap or auth object storage access
   * Maybe there are errors we are not aware that should mark as non-halt:  No space error might be the one that is fine to restart in some cases, but really not much if you have persistent volumes with certain size - how it's helpful to restart service?

Another question is why halting is harmful?

**Comment by belm0** [comment]
@bwplotka thank you for considering

Even for unrecoverable errors, users expect the server to exit.  It will create a situation of flapping restarts which will be very visible to the stock visualization and monitoring that comes with any compute hosting solution.   For example, on Google Cloud the service dashboard will clearly indicate that the server has problems and is in a backoff state.  And it is trivial to set up blanket notifications for situations like that across all hosted jobs.

It's not nice to Thanos users to expect them, from day one, to 1) know that halt state of compact needs to be monitored, and 2) set up that monitoring and alerting for it correctly.  And when they don't do this, critical errors that happen will be obscured and they won't realize it until much later.  I'll repeat our real example of this, which was a big mess:

> for several weeks we didn't notice that thanos-compact was halted due to overlapping blocks uploaded by a rouge prometheus+thanos cluster on the same replica name and store. Hundreds of overlapping blocks were added during this time.

**Exactly because thanos-compact is not in the critical path of ingestion or queries, an hence downtime is not obvious, critical errors need to be surfaced in a standard way expected by hosting solutions.**  Which is to exit on critical error.

Please change the default of debug.halt-on-error to false.

**Comment by GiedriusS** [comment]
> @bwplotka thank you for considering
> 
> Even for unrecoverable errors, users expect the server to exit. It will create a situation of flapping restarts which will be very visible to the stock visualization and monitoring that comes with any compute hosting solution. For example, on Google Cloud the service dashboard will clearly indicate that the server has problems and is in a backoff state. And it is trivial to set up blanket notifications for situations like that across all hosted jobs.
> 
> It's not nice to Thanos users to expect them, from day one, to 1) know that halt state of compact needs to be monitored, and 2) set up that monitoring and alerting for it correctly. And when they don't do this, critical errors that happen will be obscured and they won't realize it until much later. I'll repeat our real example of this, which was a big mess:
> 
> > for several weeks we didn't notice that thanos-compact was halted due to overlapping blocks uploaded by a rouge prometheus+thanos cluster on the same replica name and store. Hundreds of overlapping blocks were added during this time.
> 
> **Exactly because thanos-compact is not in the critical path of ingestion or queries-- hence downtime is not obvious-- critical errors need to be surfaced in a standard way expected by hosting solutions.** Which is to exit on critical error.
> 
> Please change the default of `debug.halt-on-error` to false.

Could you please explain what would be the difference between your proposed change and simply running Thanos Compactor on a cronjob that would exit with `1` if an error had happened? It would also 

> critical errors need to be surfaced in a standard way expected by hosting solutions

Just like you've said it would do that by using the universal Unix convention how an error gets surfaced to the user, with exit codes.

**Comment by Sayrus** [comment]
Hello,

I'd just like to add that one of our Thanos Compactor was halted for months and our dashboard was not reflecting the `thanos_compactor_halted`. We missed it and performance quickly decreased.
An Exit Code would have lead to alerts on our Kubernetes as the pod would be CrashLooping.

Now that we found this issue, we will add the debug flag as the storage we use is persistent.

> Could you please explain what would be the difference between your proposed change and simply running Thanos Compactor on a cronjob that would exit with 1 if an error had happened? It would also

This is a very selfish opinion but I think exiting on 1 is the usual behavior for most applications. Our Thanos Compactor is running as a StatefulSet and was expected to exit on 1.
The Thanos Operator deploys a Deployment so the expected pattern here is running it as an application and not a Cron. (And [they don't set the debug flag](https://github.com/banzaicloud/thanos-operator/blob/9079ed2988ff9f8f14c69be1e7aba59fafb5a3e7/pkg/resources/compactor/deployment.go))

> It's not nice to Thanos users to expect them, from day one, to 1) know that halt state of compact needs to be monitored, and 2) set up that monitoring and alerting for it correctly. And when they don't do this, critical errors that happen will be obscured and they won't realize it until much later.

Outside of "How should the application be deployed" or "What the difference with a cron would be", I think this is a good summary of the pitfall we fell in. We did not expect the application to need alerting on the `thanos_compactor_halted` metric and only discovered it while reading this issue 👍 

Thanks for documenting the whole issue, it was easy to find once we knew what was wrong.
